nncf/experimental/torch/nas/bootstrapNAS/elasticity/nas_shape_pruning_processor.py

```diff
@@ -0,0 +1,48 @@
+"""
+ Copyright (c) 2023 Intel Corporation
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+from typing import Callable, List, Optional
+
+from nncf.common.graph import NNCFGraph
+from nncf.common.graph import NNCFNode
+from nncf.common.pruning.clusterization import Clusterization
+from nncf.common.pruning.shape_pruning_processor import ShapePruningProcessor
+from nncf.common.pruning.structs import PrunedLayerInfoBase
+from nncf.common.pruning.utils import get_input_masks
+from nncf.common.tensor import NNCFTensor
+
+
+class NASShapePruningProcessor(ShapePruningProcessor):
+    """
+    BootstrapNAS shape pruning function.
+    """
+
+    def __init__(
+        self,
+        prunable_types: List[str],
+        pruning_operations_metatype: List[str],
+        get_input_masks_func: Callable[[NNCFNode, NNCFGraph], List[Optional[NNCFTensor]]] = get_input_masks,
+    ):
+        super().__init__(prunable_types, pruning_operations_metatype)
+        self.get_input_masks_func = get_input_masks_func
+
+    def _get_next_node_sparse_multiplier(
+        self, graph: NNCFGraph, next_node: NNCFNode, cluster: Clusterization[PrunedLayerInfoBase]
+    ) -> int:
```

Comment:
Could you please elaborate which exactly scenario this fix?

Reply:
This function mainly fixes the scenario where [add_dynamic_inputs](https://github.com/openvinotoolkit/nncf/blob/dd37d445e3112edcd727fcef0196bc1caede4398/nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py#L139) is used in elastic width. Some operators are only prunable in some specific cases. For security reasons, those operators are usually set as unprunable in the common function. For some specific models, in order to further improve performance, we need to add output nodes of those operators in [add_dynamic_inputs](https://github.com/openvinotoolkit/nncf/blob/dd37d445e3112edcd727fcef0196bc1caede4398/nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py#L139) to force enable their elasticity.

For example, here is part of the graph in OFA-MBV3:
![image](https://github.com/openvinotoolkit/nncf/assets/107673535/fc2f9760-b289-4eba-8d64-35b9af6d8849)
If we do not use [add_dynamic_inputs](https://github.com/openvinotoolkit/nncf/blob/dd37d445e3112edcd727fcef0196bc1caede4398/nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py#L139), node104(marked in yellow) is unprunable because the node107 and node108(marked in green) are mean type, which sets output_mask = None. However, we can force enable node104 to be prunable by adding node109 to [add_dynamic_inputs](https://github.com/openvinotoolkit/nncf/blob/dd37d445e3112edcd727fcef0196bc1caede4398/nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py#L139) and adding node104 into [overwrite_groups](https://github.com/openvinotoolkit/nncf/blob/dd37d445e3112edcd727fcef0196bc1caede4398/nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py#L137).


Compared with [get_input_masks](https://github.com/openvinotoolkit/nncf/blob/dd37d445e3112edcd727fcef0196bc1caede4398/nncf/common/pruning/utils.py#L361), `get_input_masks_with_add_dynamic_inputs` considers input masks of operators in [add_dynamic_inputs](https://github.com/openvinotoolkit/nncf/blob/dd37d445e3112edcd727fcef0196bc1caede4398/nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py#L139). Here the function `_get_next_node_sparse_multiplier` uses `get_input_masks_with_add_dynamic_inputs` instead of  [get_input_masks](https://github.com/openvinotoolkit/nncf/blob/dd37d445e3112edcd727fcef0196bc1caede4398/nncf/common/pruning/utils.py#L361) to get input masks, so it can handle the scenario with [add_dynamic_inputs](https://github.com/openvinotoolkit/nncf/blob/dd37d445e3112edcd727fcef0196bc1caede4398/nncf/experimental/torch/nas/bootstrapNAS/elasticity/elastic_width.py#L139).

Thanks!